stl/inc/memory

```diff
@@ -2416,26 +2437,298 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20
+template <class _Ty>
+struct alignas(2 * sizeof(void*)) _Atomic_ptr_base {
+protected:
+    constexpr _Atomic_ptr_base() noexcept = default;
+
+    _Atomic_ptr_base(_Ty* const _Px, _Ref_count_base* const _Ref) noexcept
+        : _Ptr(_Px), _Refptr(reinterpret_cast<const uintptr_t>(_Ref)) {}
+
+    _Ref_count_base* _LockAndLoad() const noexcept {
+        uintptr_t _Ref = _Refptr & ~1;
+        while (!_Refptr.compare_exchange_strong(_Ref, _Ref | 1)) {
+            _Ref &= ~1;
+            _YIELD_PROCESSOR();
+        }
+        return reinterpret_cast<_Ref_count_base*>(_Ref);
+    }
+
+    _Ref_count_base* _LockAndExchange(_Ref_count_base* const _Value) const noexcept {
+        uintptr_t _UValue = reinterpret_cast<uintptr_t>(_Value);
+        _UValue |= 1;
+        uintptr_t _Ref = _Refptr & ~1;
+        while (!_Refptr.compare_exchange_strong(_Ref, _UValue)) {
+            _Ref &= ~1;
+            _YIELD_PROCESSOR();
+        }
+        return reinterpret_cast<_Ref_count_base*>(_Ref);
+    }
+
+    void _Unlock() const noexcept {
+        _Refptr &= ~1;
+    }
+
+    void _StoreAndUnlock(_Ref_count_base* const _Value) const noexcept {
+        _Refptr = reinterpret_cast<uintptr_t>(_Value);
+    }
+
+    _Ty* _Ptr                         = nullptr;
+    mutable atomic<uintptr_t> _Refptr = 0;
+};
+
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_ptr_base<_Ty> {
+private:
+    using _Base = _Atomic_ptr_base<_Ty>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+    static constexpr bool is_always_lock_free = false;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return false;
+    }
+
+    void store(const shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_store_memory_order(_Order);
+        _Value._Incref();
+        _Ref_count_base* _Ref = this->_LockAndExchange(_Value._Rep);
+        if (_Ref) {
+            _Ref->_Decref();
+        }
+        this->_Ptr = _Value._Ptr;
+        this->_Unlock();
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order = memory_order::seq_cst) const noexcept {
+        _Check_load_memory_order(_Order);
+        shared_ptr<_Ty> _Result;
+        _Ref_count_base* _Ref = this->_LockAndLoad();
+        _Result._Ptr          = this->_Ptr;
+        _Result._Rep          = _Ref;
+        _Result._Incref();
+        this->_Unlock();
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(
+        const shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);

```

Comment:
This can be more efficient (1 less CAS and no refcount changes) as something like:

```
shared_ptr<_Ty> _Result;
_Result._Rep = this->_Lock_and_load();
_Result._Ptr = this->_Ptr;
this->_Ptr = _Value._Ptr;
this->_Rep.store(_Value._Rep); // also unlocks
_Value._Ptr = nullptr; // ownership of this ref has been given to this, silence decrement
_Value._Rep = nullptr;
return _Result;
```

Reply:
Done